### DISCUSSION TITLE
Nested tests?

### DISCUSSION THREAD

**Discussion by lumos42** [aptos-labs/aptos-developer-discussions#323] [Questions]
1. Is there a way to write tested tests in Aptos? Hardhat supports these types of tests.

For example:
```js
fun test() {
    a = 10;
    b = 30;
    // some state changes

    test1 {
        // retains state changes and a and b
    }

    test2 {
        // retains state changes and a and b
    }
}
```

2. Are there any setup/cleanup hooks I can use for all my tests in a particular module?

**Comment by 0x-j** [comment]
This might be what you are looking for, [you do initialization for test in one module](https://github.com/aptos-labs/aptos-names-contracts/blob/main/router/sources/router.move#L963-L966), [hen run the actual test in another module](https://github.com/aptos-labs/aptos-names-contracts/blob/main/router/sources/tests/router_tests.move)
